### ISSUE TITLE
ROBUSTNESS: Fix clang issues

### ISSUE THREAD

**Issue by HenrikBengtsson** [HenrikBengtsson/sfit#1]
``` bash
$ make clang
clang -c -pedantic -Wall -IC:/PROGRA~1/R/R-devel/include/ src/*.c
In file included from src/cfit.c:20:
In file included from src/cfit.h:22:
src/spa.h:899:25: warning: variable 'i' is uninitialized when used here
      [-Wuninitialized]
  for ( a = A[1], b = B[i], i = 1; i <= n; i++, a += m, b += m )
                        ^
src/spa.h:898:18: note: initialize the variable 'i' to silence this warning
  R *a, *b; int i, j;
                 ^
                  = 0
src/spa.h:917:25: warning: variable 'i' is uninitialized when used here
      [-Wuninitialized]
  for ( a = A[1], b = B[i], i = 1; i <= n; i++, a += m, b += m )
                        ^
src/spa.h:916:18: note: initialize the variable 'i' to silence this warning
  R *a, *b; int i, j;
                 ^
                  = 0
src/cfit.c:264:18: warning: ordered comparison between pointer and zero ('R *'
      (aka 'double *') and 'int') is an extension [-Wpedantic]
    if( Mnext[i] < 0 ) Mnext[i] = 0;
        ~~~~~~~~ ^ ~
3 warnings generated.
In file included from src/cli.c:28:
In file included from src/cfit.h:22:
src/spa.h:899:25: warning: variable 'i' is uninitialized when used here
      [-Wuninitialized]
  for ( a = A[1], b = B[i], i = 1; i <= n; i++, a += m, b += m )
                        ^
src/spa.h:898:18: note: initialize the variable 'i' to silence this warning
  R *a, *b; int i, j;
                 ^
                  = 0
src/spa.h:917:25: warning: variable 'i' is uninitialized when used here
      [-Wuninitialized]
  for ( a = A[1], b = B[i], i = 1; i <= n; i++, a += m, b += m )
                        ^
src/spa.h:916:18: note: initialize the variable 'i' to silence this warning
  R *a, *b; int i, j;
                 ^
                  = 0
src/cli.c:198:34: warning: comparison of unsigned enum expression < 0 is always
      false [-Wtautological-compare]
        if( !p || C->constraints < 0 || C->constraints > 3 )
                  ~~~~~~~~~~~~~~ ^ ~
3 warnings generated.
In file included from src/spa.c:21:
src/spa.h:899:25: warning: variable 'i' is uninitialized when used here
      [-Wuninitialized]
  for ( a = A[1], b = B[i], i = 1; i <= n; i++, a += m, b += m )
                        ^
src/spa.h:898:18: note: initialize the variable 'i' to silence this warning
  R *a, *b; int i, j;
                 ^
                  = 0
src/spa.h:917:25: warning: variable 'i' is uninitialized when used here
      [-Wuninitialized]
  for ( a = A[1], b = B[i], i = 1; i <= n; i++, a += m, b += m )
                        ^
src/spa.h:916:18: note: initialize the variable 'i' to silence this warning
  R *a, *b; int i, j;
                 ^
                  = 0
2 warnings generated.
rm -f *.o
```

**Comment by HenrikBengtsson** [comment]
Fixed the initialization issues. Remaining issues:

```
clang -c -pedantic -Wall -IC:/PROGRA~1/R/R-devel/include/ src/*.c
src/cfit.c:264:18: warning: ordered comparison between pointer and zero ('R *'
      (aka 'double *') and 'int') is an extension [-Wpedantic]
    if( Mnext[i] < 0 ) Mnext[i] = 0;
        ~~~~~~~~ ^ ~
1 warning generated.
src/cli.c:198:34: warning: comparison of unsigned enum expression < 0 is always
      false [-Wtautological-compare]
        if( !p || C->constraints < 0 || C->constraints > 3 )
                  ~~~~~~~~~~~~~~ ^ ~
1 warning generated.
```

**Comment by HenrikBengtsson** [comment]
Remaining issue:

```
clang -c -pedantic -Wall -IC:/PROGRA~1/R/R-devel/include/ src/*.c
src/cfit.c:264:18: warning: ordered comparison between pointer and zero ('R *'
      (aka 'double *') and 'int') is an extension [-Wpedantic]
    if( Mnext[i] < 0 ) Mnext[i] = 0;
        ~~~~~~~~ ^ ~
1 warning generated.
```

EDIT: _This looks like a bug_.  Should probably be:

```
    if( *Mnext[i] < 0 ) *Mnext[i] = 0;
```

This is part of the C function `cf_nn()` which is used when `C->constraints == 1` ("nonnegative").

EDIT 2: The `cfit()` function in R does _not_ pass/set/use the `-C / --contstraints` option of `cli.c` program.  The default is `C->constraints = 0` ("none").  Thus, the conclusion is that this bug has not/does not/will not affect the `cfit()` function in R.

**Comment by HenrikBengtsson** [comment]
Ok, more on this, the `R` type in:

``` sh
clang -c -pedantic -Wall -IC:/PROGRA~1/R/R-devel/include/ src/*.c
src/cfit.c:264:18: warning: ordered comparison between pointer and zero ('R *' (aka 'double *') and 'int') is an extension [-Wpedantic]
    if( Mnext[i] < 0 ) Mnext[i] = 0;
        ~~~~~~~~ ^ ~
```

is `double` according to the **warning** and is defined in `src/spa.h` (there is can also be defined as `float` if macro `R_IS_FLOAT` is set, which it isn't).  Either way, we can guide the compiler to use `0.0` instead of `0` to tell it is a floating point constant rather than an integer.  However, if we do that, we all of a sudden get **errors**:

``` sh
clang -c -pedantic -Wall -IC:/PROGRA~1/R/R-devel/include/ src/*.c
src/cfit.c:264:18: error: invalid operands to binary expression ('R *' (aka 'double *') and 'double')
    if( Mnext[i] < 0.0 ) Mnext[i] = 0.0;
        ~~~~~~~~ ^ ~~~
src/cfit.c:264:35: error: assigning to 'R *' (aka 'double *') from incompatible
      type 'double'
    if( Mnext[i] < 0.0 ) Mnext[i] = 0.0;
                                  ^ ~~~
```

This is not surprising at all, because now it's unambiguous to the compiler what we're trying to do; we're no longer trying to assign _pointer_  `Mnext[i]` of type `R *` and integer value, which is kind of alright with pointers, but a double, which is certainly not valid.

This is in full agreement with the observation that we should really use:

``` c
void cf_nn ( int k, int m, R **M, R **Mnext, void *param)
{
  int i;
  for ( i = 1; i <= k*m; i++ )
    if( *Mnext[i] < 0.0 ) *Mnext[i] = 0.0;
}
```

and if we do, the compile errors go away.

Also, since `Mnext[i]` is a pointer and therefore holds an address in memory it is/should be an _unsigned_ address integer.  This means that the condition `Mnext[i] < 0` if this code would have been called, would always be FALSE, which means that the assignment `Mnext[i] = 0` would never take place, which would be problematic, because it would set address pointer to zero.  So, this bug/typo has been harmless.  The only result is that `cf_nn()` wouldn't do anything at all, which in turn would equal the `"none"` (==0) constraints case.

**Comment by HenrikBengtsson** [comment]
Fixed in Commit 05aee73.

**Comment by HenrikBengtsson** [comment]
The `covr` run [results](https://coveralls.io/files/448765217) show that at least the package tests do not reach the `cf_nn()` function that was buggy.
